mlir_graphblas/mlir_builder.py

```diff
@@ -7,54 +7,125 @@
 from .sparse_utils import MLIRSparseTensor
 from .functions import BaseFunction
 from .engine import parse_mlir_string
 
-from typing import Dict, List, Tuple
+from typing import Dict, List, Tuple, Sequence, Generator, Optional, Union, Iterable
 
 #############
 # IR Builer #
 #############
 
+def _canonicalize_mlir_type_string(type_string: str) -> str:
+    '''Canonicalize by round-tripping through PyMLIR.'''
+    return mlir.parse_string(f'!dummy_alias = type {type_string}').body[0].value.dump()
+
+def mlir_type_strings_equal(type_1: str, type_2: str) -> bool:
+    # The type strings can vacuously differ (e.g. in white space).
+    # Thus, we must deterministically canonicalize the type strings.
+    if type_1 == type_2:
+        return True
+    if _canonicalize_mlir_type_string(type_1) == _canonicalize_mlir_type_string(type_2):
+        return True
+    return False
 
 class MLIRVar:
 
     """Input variables to be used by MLIRFunctionBuilder."""
-
-    def __init__(self, var_name: str, var_type: str) -> None:
-        """var_type is the type as represented in MLIR code."""
+    
+    def __init__(self, var_name: str, *var_types: str) -> None:
+        """Elements of var_types are types as represented in MLIR code."""
+        assert len(var_types) > 0
         self.var_name = var_name
-        self.var_type = var_type
+        self.var_types = var_types
         return
 
+    def __eq__(self, other: 'MLIRVar') -> bool:
+        return (
+            isinstance(other, MLIRVar) 
+            and self.var_name == other.var_name 
+            and all(
+                mlir_type_strings_equal(*type_pair)
+                for type_pair in zip(self.var_types, other.var_types)
+            )
+        )
+
+    @property
+    def num_values(self) -> int:
+        return len(self.var_types)
+
     @property
-    def var_string(self) -> str:
-        return f"%{self.var_name}"
+    def var_type(self) -> str:
+        if not len(self.var_types) == 1:
+            raise TypeError(f"{self} has multiple types.")
+        var_type, = self.var_types
+        return var_type
+
+    def __repr__(self) -> str:
+        attributes_string = ', '.join(f'{k}={repr(self.__dict__[k])}' for k in sorted(self.__dict__.keys()))
+        return f'{self.__class__.__name__}({attributes_string})'
+    
+    def assign_string(self) -> str:
+        # TODO specify number of expected returned values as arg and sanity check
+        answer = f"%{self.var_name}"
+        if self.num_values != 1:
+            answer += f":{self.num_values}"
+        return answer
+
+    def access_string(self, index: Optional[int] = None) -> str:
+        answer = f"%{self.var_name}"
+        if index is None:
+            if self.num_values != 1:
+                raise RuntimeError("Attempting to use variable containing multiple values as a singleton variable.")
+        elif not isinstance(index, int):
+            raise LookupError(f"{self.__class__} only supports integer indexing.")
```

Comment:
This should be changed to a `TypeError` to be analogous to this:
```

>>> [1,2,3]['bad_index']
<stdin>:1: SyntaxWarning: list indices must be integers or slices, not str; perhaps you missed a comma?
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: list indices must be integers or slices, not str
```

Reply:
https://github.com/metagraph-dev/mlir-graphblas/pull/47/commits/dfe7d563acf99e1608e2ceb6967e830ce67dc3ce implements these changes.